agent/consul/gateways/controller_gateways.go

```diff
@@ -2,33 +2,202 @@ package gateways
+		return nil
+	}
+
+	r.ensureBoundGateway(metaGateway)
+
+	r.logger.Debug("started reconciling gateway")
+
+	routes, err := r.retrieveAllRoutesFromStore()
+
+	if err != nil {
+		return err
+	}
+
+	boundGateways, routeErrors := BindRoutesToGateways([]*gatewayMeta{metaGateway}, routes...)
+
+	//In this loop there should only be 1 bound gateway returned, but looping over all returned gateways
+	//to make sure nothing gets dropped and handle case where 0 gateways are returned
+	for _, boundGateway := range boundGateways {
+		// now update the gateway state
+		r.logger.Debug("persisting gateway state", "state", boundGateway)
+		if err := r.store.Update(boundGateway); err != nil {
+			msg := "error persisting state"
+			r.logger.Error(msg, "error", err)
+			return errors.Wrap(err, msg)
+		}
+
+		// then update the gateway status
+		r.logger.Debug("persisting gateway status", "gateway", metaGateway.Gateway)
+		if err := r.store.UpdateStatus(metaGateway.Gateway, err); err != nil {
+			return err
+		}
+	}
+
+	//// and update the route statuses
+	for route, routeError := range routeErrors {
+		configEntry := r.resourceReferenceToBoundRoute(route)
+		r.logger.Error("route binding error:", routeError)
+		if err := r.store.UpdateStatus(configEntry, routeError); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *apiGatewayReconciler) retrieveAllRoutesFromStore() ([]structs.BoundRoute, error) {
+	tcpRoutes, err := r.store.GetConfigEntriesByKind(structs.TCPRoute)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO not implemented
+	//httpRoutes, err := r.store.GetConfigEntriesByKind(structs.HTTPRoute)
+	//if err != nil {
+	//	return nil, err
+	//}
+
+	routes := []structs.BoundRoute{}
+	for _, r := range tcpRoutes {
+		if r == nil {
+			continue
+		}
+		routes = append(routes, r.(*structs.TCPRouteConfigEntry))
+	}
+	//TODO not implemented
+	//for _, r := range httpRoutes {
+	//	routes = append(routes, r.(*structs.HTTPRouteConfigEntry))
+	//}
+	return routes, nil
+}
+
+func (r *apiGatewayReconciler) initGatewayMeta(req controller.Request) (*gatewayMeta, error) {
+	metaGateway := &gatewayMeta{}
+
+	apiGateway, err := r.store.GetConfigEntry(req.Kind, req.Name, req.Meta)
+	if err != nil {
+		return nil, err
+	}
+
+	if apiGateway == nil {
+		//gateway needs to be deleted, don't init
+		return nil, nil
+	}
+
+	metaGateway.Gateway = apiGateway.(*structs.APIGatewayConfigEntry)
+
+	boundGateway, err := r.store.GetConfigEntry(structs.BoundAPIGateway, req.Name, req.Meta)
+	if err != nil {
+		return nil, err
+	}
+
+	//initialize object, values get copied over in ensureBoundGateway if they don't exist
+	metaGateway.BoundGateway = boundGateway.(*structs.BoundAPIGatewayConfigEntry)
+	return metaGateway, nil
+}
+
+func (r *apiGatewayReconciler) resourceReferenceToBoundRoute(ref structs.ResourceReference) structs.ControlledConfigEntry {
+	//TODO currently have to retrieve from the store to persist parent field on update call, is there a better way to do this?
```

Comment:
Not quite sure what you mean here. I'll check in with you on a video chat today

Reply:
Available to talk through this whenever, but my concern was if I were to update the status with a niled out parent field it would nil out the parent field in the store as well.

Reply:
We chatted and the gist is:
1. Receive reference to config entry, we need to update the status
2. Update the status
3. Write config entry w/ status to the store

The reason we fetch the config entry here is so as not to wipe out the config entry contents when trying to write the status